src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -963,6 +974,19 @@ ExecuteModifyTasks(List *taskList, bool expectResults, ParamListInfo paramListIn
 		return 0;
 	}
 
+	foreach(taskCell, taskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		ShardInterval *shardInterval = LoadShardInterval(task->anchorShardId);
+		Oid relationId = shardInterval->relationId;
+
+		/*
+		 * If we are dealing with partitioned table, we also need to lock its
+		 * partitions.
+		 */
+		LockPartitionRelationsIfPartitioned(relationId, RowExclusiveLock);
```

Comment:
I think it would make more sense if the `if (PartitionTable(...)` statement was around the loop and the function was called `LockPartitionRelations`. The same goes for `LockParentRelationShardResources`. Now we do a bunch of work for now reason when we're dealing with a non-partitioned table.